gix-worktree/src/diff.rs

```diff
@@ -0,0 +1,292 @@
+        &mut self,
+        worktree_path: &Path,
+        index_entry: &index::Entry,
+        buf: &mut Vec<u8>,
+        capabilities: &fs::Capabilities,
+    ) -> Result<(), read::Error> {
+        if self.mode_change.is_some() || !self.stat_changed || self.data_changed {
+            return Ok(());
+        }
+        let data = read_blob_to_buf_with_meta(
+            worktree_path,
+            index_entry.mode.contains(index::entry::Mode::SYMLINK),
+            buf,
+            capabilities,
+        )?;
+        let header = loose_header(gix_object::Kind::Blob, data.len());
+        let hash_changed = match index_entry.id {
+            ObjectId::Sha1(entry_hash) => {
+                let mut file_hash = hash::Sha1::default();
+                file_hash.update(&header);
+                file_hash.update(&data);
+                let file_hash = file_hash.digest();
+                entry_hash != file_hash
+            }
+        };
+        self.data_changed = hash_changed;
+        Ok(())
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Could not convert path to UTF8 {path}")]
+    IllformedUtf8 { path: BString },
+    #[error("The clock was off when reading file related metadata after updating a file on disk")]
+    Time(#[from] std::time::SystemTimeError),
+    #[error("IO error while writing blob or reading file metadata or changing filetype")]
+    Io(#[from] io::Error),
+}
+
+#[derive(Clone, Debug)]
+/// A change between the index and the worktree computed by [`compate_to_index`]
+pub struct Change<'a> {
+    /// The index entry that changed
+    pub index_entry: &'a index::Entry,
+    /// The on-disk worktree path corresponding to this entry
+    pub worktree_path: PathBuf,
+    /// How this index entry changed
+    pub kind: ChangeKind,
+    /// file metadata that can be reused (optimization)
+    pub fstat: Option<std::fs::Metadata>,
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
+///
+pub enum ChangeKind {
+    /// An index entry has no corresponding file in the worktree
+    Removed,
+    /// Ar new files that has been marked with git add but has not yet been
+    /// checked in yet. No diff is computed for these files because whatever is
+    /// on disk at commit time will be used
+    Added,
+    /// Called for files that may have changed in some form as indicated by `change`.
+    /// Note that this doesn't necessarily mean that the *content* of the file changed
+    /// see [`FileStatus`] for details
+    Modified {
+        /// How the file was modified exactly
+        modification: FileModification,
+        /// Whether this (changed) file also has an unresolved merge conflict
+        conflict: bool,
+    },
+    /// There are unresolved merge conflicts for this file
+    /// but it has not changed on disk
+    Conflict,
+}
+
+/// Computes the changes between the index and the worktree
+pub fn compare_to_index<'a: 'b, 'b>(
```

Comment:
What really helps me is to name these things `changes_to_obtain_X`, so it becomes clear if this is `worktree_changes_to_obtain_index` or `index_changes_to_obtain_worktree`. Right now it's `gix_worktree::diff::compare_to_index` which makes me think that it's the changes the worktree has to make to become the index, but I think it's the opposite. `ChangeKind` can be redefined to be 'the change X needs to become Y`. Maybe double-check, but this is how it's done in `gix-diff` as well.